File: src/main/java/com/xxxt/cupm/ClientModInit.java
package com.xxxt.cupm;

import net.minecraft.client.multiplayer.ClientLevel;
import net.minecraft.client.renderer.item.ClampedItemPropertyFunction;
import net.minecraft.client.renderer.item.ItemProperties;
import net.minecraft.world.entity.LivingEntity;
import net.minecraft.world.item.Item;
import net.minecraft.world.item.ItemStack;
import net.neoforged.api.distmarker.Dist;
import net.neoforged.bus.api.SubscribeEvent;
import net.neoforged.fml.common.EventBusSubscriber;
import net.neoforged.neoforge.client.event.RegisterClientReloadListenersEvent;
import org.jetbrains.annotations.NotNull;

import java.util.Arrays;
import java.util.List;

@EventBusSubscriber(modid = CUPMMod.MODID, bus = EventBusSubscriber.Bus.MOD, value = Dist.CLIENT)
public class ClientModInit {

    @SubscribeEvent
    public static void onClientSetup(RegisterClientReloadListenersEvent event) {
        List<Item> shiftCaps = Arrays.asList(
                CUPMItems.INSTANCE.getSILVER_CAP().get(),
                CUPMItems.INSTANCE.getCOPPER_CAP().get(),
                CUPMItems.INSTANCE.getIRON_CAP().get(),
                CUPMItems.INSTANCE.getOBSIDIAN_CAP().get()
        );

        for (Item cap : shiftCaps) {
            ItemProperties.register(cap, CUPMTags.INSTANCE.getCAP_STATE_TAG().getId(), new ClampedItemPropertyFunction() {
                @Override
                public float unclampedCall(@NotNull ItemStack stack, ClientLevel level, LivingEntity entity, int seed) {
                    return stack.getComponents().getOrDefault(CUPMTags.INSTANCE.getCAP_STATE_TAG().get(), 0) * 0.1f;
                }
            });
        }
    }
}

File: src/main/kotlin/com/xxxt/cupm/events/ClientModInit.kt
package com.xxxt.cupm.events

import com.xxxt.cupm.CUPMItems
import com.xxxt.cupm.CUPMMod
import com.xxxt.cupm.CUPMTags
import net.minecraft.client.multiplayer.ClientLevel
import net.minecraft.client.renderer.item.ClampedItemPropertyFunction
import net.minecraft.client.renderer.item.ItemProperties
import net.minecraft.world.entity.LivingEntity
import net.minecraft.world.item.ItemStack
import net.neoforged.api.distmarker.Dist
import net.neoforged.bus.api.SubscribeEvent
import net.neoforged.fml.common.EventBusSubscriber
import net.neoforged.neoforge.client.event.RegisterClientReloadListenersEvent

@EventBusSubscriber(modid = CUPMMod.Companion.MODID, bus = EventBusSubscriber.Bus.MOD, value = [Dist.CLIENT])
object ClientModInit {
    @SubscribeEvent
    @JvmStatic
    fun onClientSetup(event: RegisterClientReloadListenersEvent) {
        val shiftCaps = listOf(
            CUPMItems.SILVER_CAP.get(),
            CUPMItems.COPPER_CAP.get(),
            CUPMItems.IRON_CAP.get(),
            CUPMItems.OBSIDIAN_CAP.get()
        )

        for (cap in shiftCaps) {
            ItemProperties.register(cap, CUPMTags.CAP_STATE_TAG.id,
                object : ClampedItemPropertyFunction {
                override fun unclampedCall(
                    stack: ItemStack,
                    level: ClientLevel?,
                    entity: LivingEntity?,
                    seed: Int
                ): Float {
                    return stack.getComponents().getOrDefault(CUPMTags.CAP_STATE_TAG.get(), 0) * 0.1f
                }
            })
        }
    }
}